src/lib.rs

```diff
@@ -0,0 +1,408 @@
+        }
+    }
+
+    /// Generate the rustdoc file from the project root directory, i.e. the directory containing the crate source.
+    /// It can be a workspace or a single package.
+    /// Same as `from_git_revision`, but with the current git revision.
+    pub fn from_root(project_root: impl Into<PathBuf>) -> Self {
+        Self {
+            source: RustdocSource::Root(project_root.into()),
+        }
+    }
+
+    /// Generate the rustdoc file from the project at a given git revision.
+    pub fn from_git_revision(
+        project_root: impl Into<PathBuf>,
+        revision: impl Into<String>,
+    ) -> Self {
+        Self {
+            source: RustdocSource::Revision(project_root.into(), revision.into()),
+        }
+    }
+
+    /// Generate the rustdoc file from the largest-numbered non-yanked non-prerelease version
+    /// published to the cargo registry. If no such version, uses
+    /// the largest-numbered version including yanked and prerelease versions.
+    pub fn from_latest_version() -> Self {
+        Self {
+            source: RustdocSource::Version(None),
+        }
+    }
+
+    pub fn from_version(version: impl Into<String>) -> Self {
+        Self {
+            source: RustdocSource::Version(Some(version.into())),
+        }
+    }
+}
+
+enum RustdocSource {
+    /// Path to the Rustdoc json file. Use this option when you have already generated the rustdoc file.
+    Rustdoc(PathBuf),
+    /// Project root directory, i.e. the directory containing the crate source.
+    /// It can be a workspace or a single package.
+    Root(PathBuf),
+    /// Project root directory and Git Revision.
+    Revision(PathBuf, String),
+    /// Version from cargo registry to lookup. E.g. "1.0.0".
+    /// If `None`, uses the largest-numbered non-yanked non-prerelease version
+    /// published to the cargo registry. If no such version, uses
+    /// the largest-numbered version including yanked and prerelease versions.
+    Version(Option<String>),
+}
+
+/// Which packages to analyze.
+#[derive(Default)]
+struct Scope {
+    mode: ScopeMode,
+}
+
+enum ScopeMode {
+    /// All packages except the excluded ones.
+    DenyList(PackageSelection),
+    /// Packages to process (see `cargo help pkgid`)
+    AllowList(Vec<String>),
+}
+
+impl Default for ScopeMode {
+    fn default() -> Self {
+        Self::DenyList(PackageSelection::default())
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct PackageSelection {
+    selection: ScopeSelection,
+    excluded_packages: Vec<String>,
+}
+
+impl PackageSelection {
+    pub fn new(selection: ScopeSelection) -> Self {
+        Self {
+            selection,
+            excluded_packages: vec![],
+        }
+    }
+
+    pub fn with_excluded_packages(&mut self, packages: Vec<String>) -> &mut Self {
+        self.excluded_packages = packages;
+        self
+    }
+}
+
+#[derive(Default, PartialEq, Eq, Clone)]
+pub enum ScopeSelection {
```

Comment:
```suggestion
#[non_exhaustive]
#[derive(Debug, Default, PartialEq, Eq, Clone)]
pub enum ScopeSelection {
```
Unless there's a reason not to?

Reply:
done in https://github.com/obi1kenobi/cargo-semver-checks/pull/314/commits/f626d7ebaed9becbf30edaaf24ff273a75fa0830